llm/data_gemma/base.py

```diff
@@ -0,0 +1,186 @@
+
+  # For point: val and date is set
+  val: str = ''
+  date: str = ''
+  # For table: table is set
+  table: str = ''
+
+  unit: str = ''
+  title: str = ''
+  src: str = ''
+  url: str = ''
+  var: str = ''
+  score: float = 0.0
+
+  # The original LLM Value in case of RIG.
+  llm_val: str = ''
+
+  def footnote(self) -> str:
+    return (
+        f'Per {self.src}, value was {self.val}{self._dunit()} in {self.date}.'
+        f' See more at {self.url}'
+    )
+
+  def debug(self) -> str:
+    if not self.title:
+      return ''
+    if self.table:
+      return self.answer()
+    return (
+        f'"{self.title}" was {self.val}{self._dunit()} in'
+        f' {self.date} per {self.src} ({self.var}:{self.score})'
+    )
+
+  def answer(self) -> str:
+    if self.table:
+      return f'{self.header()}\n{self.table}'
+    else:
+      return (
+          f'According to {self.src}, "{self.title}" was'
+          f' {self.val}{self._dunit()} in {self.date}.'
+      )
+
+  def header(self) -> str:
+    if self.table:
+      if self.unit:
+        header = f'{self.title} (unit: {self.unit})'
+      else:
+        header = f'{self.title}'
+      return f'{header}, according to {self.src}'
+
+    return self.title
+
+  def val_and_unit(self) -> str:
+    return f'{self.val}{self._dunit()}'
+
+  def _dunit(self) -> str:
+    return ' ' + self.unit if self.unit else ''
+
+
+@dataclasses.dataclass(frozen=True)
+class FlowResponse:
+  """A response from Flow."""
+
+  main_text: str = ''
+  footnotes: str = ''
+  tables_str: str = ''
+
+  llm_calls: list[LLMCall] = dataclasses.field(default_factory=list)
+  dc_calls: list[DCCall] = dataclasses.field(default_factory=list)
+  dc_duration_secs: float = 0.0
+
+  def duration_secs(self) -> float:
+    return (
+        sum([r.duration_secs for r in self.llm_calls]) + self.dc_duration_secs
+    )
+
+  def answer(self, include_aux: bool = True) -> str:
+    """Returns a string representation of the response."""
+
+    lines = []
+    lines.append(self.main_text)
+
+    if include_aux and self.footnotes:
+      lines.append('\n#### FOOTNOTES ####')
+      lines.append(self.footnotes)
+
+    if include_aux and self.tables_str:
+      lines.append('\n#### TABLES ####')
+      lines.append(self.tables_str)
+
+    return '\n'.join(lines)
+
+  def debug(self) -> str:
+    """Returns a string representation of the response."""
+
+    lines = []
+    lines.append('\n\n## LLM INTERACTIONS ##\n')
+    for i, llm_response in enumerate(self.llm_calls):
+      lines.append(llm_response.debug(i))
+
+    lines.append('\n\n## DC INTERACTIONS ##\n')
+    for dc_response in self.dc_calls:
+      dbg = dc_response.debug()
+      if dbg:
+        lines.append(dbg)
+    lines.append(f'\n\n## DC Duration {self.dc_duration_secs} ##')
+    lines.append(f'\n\n## Total Duration {self.duration_secs()} ##')
+
+    return '\n'.join(lines)
+
+  def json(self) -> dict[str, Any]:
+    return dataclasses.asdict(self)
+
+
+class LLM(Protocol):
```

Comment:
Maybe make LLM more specific?

Reply:
I can't come up with a better name for the base class.  Any ideas?